File: buzu/src/test/java/br/net/buzu/metadata/AbstractMetadataTest.java
package br.net.buzu.metadata;

import br.net.buzu.pplspec.model.*;
import br.net.buzu.util.StaticBehave;
import org.junit.Test;

import java.util.List;

import static br.net.buzu.pplspec.model.DomainKt.domainOf;
import static org.junit.Assert.*;

/**
 * @author Douglas Siviotti
 *
 */
public class AbstractMetadataTest {

	private static Metadata createSample(String name, Subtype subtype, int size, int minOccurs, int maxOccurs) {
		Domain domain = domainOf("white", "black", "red");
		MetaInfo metaInfo = new MetaInfo("", 31, name, subtype, size, 0, minOccurs, maxOccurs, domain, "red", "XYZ");
		return new GenericSimpleMetadata(Kind.ARRAY, metaInfo);

	}

	@Test(expected = NullPointerException.class)
	public void testNoKind() {
		new GenericSimpleMetadata(null, null);
	}

	@Test(expected = NullPointerException.class)
	public void testNoMetainfo() {
		new GenericSimpleMetadata(Kind.ARRAY, null);
	}

	@Test
	public void test() {
		Domain domain = domainOf("Diamons", "white", "black", "red");
		MetaInfo metaInfo = new MetaInfo("", 31, "test", Subtype.STRING, 10, 0, 1, 2, domain, "red", "XYZ");

		Metadata m1 = new GenericSimpleMetadata(Kind.RECORD, metaInfo);

		assertEquals("test", m1.name());
		assertEquals(Subtype.STRING, m1.info().getSubtype());
		assertEquals(10, m1.info().getSize()); // override MetaInfo
		assertEquals(0, m1.info().getScale());
		assertEquals(1, m1.info().getMinOccurs());
		assertEquals(true, m1.info().isRequired());
		assertEquals(2, m1.info().getMaxOccurs()); // override MetaInfo
	}

	@Test
	public void testEqualsHashcodeToString() {
		Metadata m1 = createSample("test", Subtype.STRING, 10, 1, 5);
		Metadata m2 = createSample("test", Subtype.STRING, 10, 1, 5);

		// assertEquals(m1, m2);
		assertTrue(m1.equals(m2));
		assertTrue(m1.hashCode() == m2.hashCode());
		assertTrue(m1.toString().equals(m2.toString()));
		assertTrue(m2.equals(m1));
		assertTrue(m1.equals(m1));
		assertFalse(m1.equals(null));
	}

}

class GenericSimpleMetadata extends BasicMetadata {

	public GenericSimpleMetadata(Kind kind, MetaInfo metaInfo) {
		super(metaInfo);
	}

	@SuppressWarnings("unchecked")
	@Override
	public List<BasicMetadata> children() {
		// TODO Auto-generated method stub
		return null;
	}

	@Override
	public boolean hasChildren() {
		// TODO Auto-generated method stub
		return false;
	}


	@Override
	public boolean isStatic() {
		return StaticBehave.isStatic(this);
	}

}

File: ppl-lib/src/test/kotlin/br/net/buzu/pplimpl/metadata/AbstractMetadataTest.kt
package br.net.buzu.pplimpl.metadata

import br.net.buzu.pplspec.model.*
import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.Test

/**
 * @author Douglas Siviotti
 */
class AbstractMetadataTest {

    private fun createSample(name: String, subtype: Subtype, size: Int, minOccurs: Int, maxOccurs: Int): Metadata {
        val domain = domainOf("white", "black", "red")
        val metaInfo = MetaInfo(31, name, subtype, size, 0, minOccurs, maxOccurs, domain, "red", "XYZ")
        return GenericSimpleMetadata(Kind.ARRAY, metaInfo)

    }

    @Test
    fun test() {
        val domain = domainOf("Diamons", "white", "black", "red")
        val metaInfo = MetaInfo(31, "test", Subtype.STRING, 10, 0, 1, 2, domain, "red", "XYZ")

        val m1 = GenericSimpleMetadata(Kind.RECORD, metaInfo)

        assertEquals("test", m1.name())
        assertEquals(Subtype.STRING, m1.info().subtype)
        assertEquals(10, m1.info().size) // override MetaInfo
        assertEquals(0, m1.info().scale)
        assertEquals(1, m1.info().minOccurs)
        assertEquals(true, m1.info().isRequired)
        assertEquals(2, m1.info().maxOccurs) // override MetaInfo
    }

    @Test
    fun testEqualsHashcodeToString() {
        val m1 = createSample("test", Subtype.STRING, 10, 1, 5)
        val m2 = createSample("test", Subtype.STRING, 10, 1, 5)

        // assertEquals(m1, m2);
        assertTrue(m1 == m2)
        assertTrue(m1.hashCode() == m2.hashCode())
        assertTrue(m1.toString() == m2.toString())
        assertTrue(m2 == m1)
        assertTrue(m1 == m1)
        assertFalse(m1 == null)
    }

}

internal class GenericSimpleMetadata(kind: Kind, metaInfo: MetaInfo) : BasicMetadata(metaInfo) {

    override fun <T: Metadata> children(): List<T> {
        return listOf()
    }

    override fun hasChildren(): Boolean {
        // TODO Auto-generated method stub
        return false
    }


    override fun isStatic(): Boolean {
        return false
    }

}
